### PR TITLE
Add TiFlash support

### PR DESCRIPTION
**Pull request by birdstorm**
TODO:

- [x] 支持在topology中定义tiflash的配置参数 
- [x] 自定义端口
- [x] 覆盖默认参数
- [ ] 如果集群中部署了tifalsh，自动设置pd的配置replication.enable-placement-rules = true
- [ ] 如果tikv的节点数据小于3，自动设置pd的参数 replication.max-replicas: 1
- [ ] 多盘支持
- [ ] prometheus & grafana

文档：https://docs.google.com/document/d/1RqmXsmmYXQEj_qV8HcmguLK6GJY0L3PZF-ogq02EdR8/edit#

### CODE REVIEW

File: pkg/ansible/dirs.go [hunk 1]
```diff
@@ -114,6 +114,26 @@ func parseDirs(host *aini.Host, ins meta.InstanceSpec) (meta.InstanceSpec, error
...
+	case meta.ComponentTiFlash:
+		serviceFile := fmt.Sprintf("%s/%s-%d.service",
+			systemdUnitPath,
+			meta.ComponentTiFlash,
+			ins.GetMainPort())
+		cmd := fmt.Sprintf("cat `grep 'ExecStart' %s | sed 's/ExecStart=//'`", serviceFile)
+		stdout, _, err := e.Execute(cmd, false)
+		if err != nil {
+			return ins, nil
+		}
```
Comment (lonng) [inline pkg/ansible/dirs.go:126]: This part has been covered in L43.

File: pkg/meta/logic.go [hunk 2]
```diff
@@ -163,6 +165,22 @@ func (i *instance) mergeServerConfig(e executor.TiOpsExecutor, globalConf, insta
 	return e.Transfer(fp, dst, false)
 }
 
+// mergeServerConfig merges the server configuration and overwrite the global configuration
+func (i *instance) mergeTiFlashLearnerServerConfig(e executor.TiOpsExecutor, globalConf, instanceConf map[string]interface{}, paths DirPaths) error {
+	fp := filepath.Join(paths.Cache, fmt.Sprintf("%s-learner_%s-%d.toml", i.ComponentName(), i.GetHost(), i.GetPort()))
```
Comment (lonng) [inline pkg/meta/logic.go:170]: Prefer
```suggestion
	fp := filepath.Join(paths.Cache, fmt.Sprintf("%s-learner-%s-%d.toml", i.ComponentName(), i.GetHost(), i.GetPort()))
```

File: pkg/meta/logic.go [hunk 3]
```diff
@@ -526,6 +544,219 @@ func (i *PDInstance) ScaleConfig(e executor.TiOpsExecutor, b *Specification, clu
...
+    users.default.networks.ip: "::/0"
+    users.readonly.password: ""
+    users.readonly.profile: "readonly"
+    users.readonly.quota: "default"
+    users.readonly.networks.ip: "::/0"
+    profiles.default.load_balancing: "random"
+    profiles.default.max_memory_usage: 10000000000
+    profiles.default.use_uncompressed_cache: 0
+    profiles.readonly.readonly: 1
+`, cfg.DeployDir, cfg.DeployDir, cfg.TCPPort, cfg.HTTPPort,
```
Comment (lonng) [inline pkg/meta/logic.go:639]: There are many duplicated variables, you can simply the format string
```yaml
server_configs:
  tiflash:
    default_profile: "default"
    display_name: "TiFlash"
    listen_host: "0.0.0.0"
    mark_cache_size: 5368709120
    tmp_path: "%[1]s/tiflash/data/tmp"
    path: "%[1]s/tiflash/data/db"
    tcp_port: %d
    http_port: %d
    flash.tidb_status_addr: "%s"
    flash.service_addr: "%s:%d"
    flash.flash_cluster.cluster_manager_path: "%s/bin/tiflash/flash_cluster_manager"
    flash.flash_cluster.log: "%s/log/tiflash_cluster_manager.log"
    flash.flash_cluster.master_ttl: 60
    flash.flash_cluster.refresh_interval: 20
    flash.flash_cluster.update_rule_interval: 5
    flash.proxy.config: "%s/conf/tiflash-learner.toml"
    status.metrics_port: %d
    logger.errorlog: "%s/log/tiflash_error.log"
    logger.log: "%s/log/tiflash.log"
    logger.count: 20
    logger.level: "debug"
    logger.size: "1000M"
    application.runAsDaemon: true
    raft.pd_addr: "%s"
    quotas.default.interval.duration: 3600
    quotas.default.interval.errors: 0
    quotas.default.interval.execution_time: 0
    quotas.default.interval.queries: 0
    quotas.default.interval.read_rows: 0
    quotas.default.interval.result_rows: 0
    users.default.password: ""
    users.default.profile: "default"
    users.default.quota: "default"
    users.default.networks.ip: "::/0"
    users.readonly.password: ""
    users.readonly.profile: "readonly"
    users.readonly.quota: "default"
    users.readonly.networks.ip: "::/0"
    profiles.default.load_balancing: "random"
    profiles.default.max_memory_usage: 10000000000
    profiles.default.use_uncompressed_cache: 0
    profiles.readonly.readonly: 1
```
to make it easier to read.

File: pkg/meta/logic.go [hunk 4]
```diff
@@ -526,6 +544,219 @@ func (i *PDInstance) ScaleConfig(e executor.TiOpsExecutor, b *Specification, clu
...
+    server.engine-addr: "%s:%d"
+    server.addr: "0.0.0.0:%d"
+    server.advertise-addr: "%s:%d"
+    server.status-addr: "%s:%d"
+    storage.data-dir: "%s/tiflash/data/tiflash"
+    rocksdb.wal-dir: ""
+    security.ca-path: ""
+    security.cert-path: ""
+    security.key-path: ""
+`, cfg.DeployDir, cfg.IP, cfg.FlashServicePort, cfg.FlashProxyPort,
```
Comment (lonng) [inline pkg/meta/logic.go:673]: ditto

File: pkg/meta/topology.go [hunk 5]
```diff
@@ -281,6 +284,51 @@ func (s PDSpec) IsImported() bool {
...
+
+// Role returns the component role of the instance
+func (s TiFlashSpec) Role() string {
+	return ComponentTiFlash
+}
+
+// SSH returns the host and SSH port of the instance
+func (s TiFlashSpec) SSH() (string, int) {
+	return s.Host, s.SSHPort
+}
```
Comment (lonng) [inline pkg/meta/topology.go:320]: L313-L320 doesn't need anymore.

File: pkg/ansible/dirs.go (hunk 1 above)
Comment (AstroProfundis) [reply pkg/ansible/dirs.go:126]: The `data_dir` and `log_dir` are set in config file, so we should add a process to parse them in the .toml config file.

File: pkg/meta/logic.go [hunk 6]
```diff
@@ -526,6 +544,219 @@ func (i *PDInstance) ScaleConfig(e executor.TiOpsExecutor, b *Specification, clu
...
+	topo := TopologySpecification{}
+
+	err := yaml.Unmarshal([]byte(fmt.Sprintf(`
+server_configs:
+  tiflash:
+    default_profile: "default"
+    display_name: "TiFlash"
+    listen_host: "0.0.0.0"
+    mark_cache_size: 5368709120
+    tmp_path: "%s/tiflash/data/tmp"
```
Comment (AstroProfundis) [inline pkg/meta/logic.go:601]: It's better use `data_dir` here.

File: pkg/meta/logic.go [hunk 7]
```diff
@@ -526,6 +544,219 @@ func (i *PDInstance) ScaleConfig(e executor.TiOpsExecutor, b *Specification, clu
...
+    path: "%s/tiflash/data/db"
+    tcp_port: %d
+    http_port: %d
+    flash.tidb_status_addr: "%s"
+    flash.service_addr: "%s:%d"
+    flash.flash_cluster.cluster_manager_path: "%s/bin/tiflash/flash_cluster_manager"
+    flash.flash_cluster.log: "%s/log/tiflash_cluster_manager.log"
```
Comment (AstroProfundis) [inline pkg/meta/logic.go:608]: It's better use `log_dir` here.

File: pkg/meta/logic.go [hunk 8]
```diff
@@ -526,6 +544,219 @@ func (i *PDInstance) ScaleConfig(e executor.TiOpsExecutor, b *Specification, clu
...
+    flash.flash_cluster.master_ttl: 60
+    flash.flash_cluster.refresh_interval: 20
+    flash.flash_cluster.update_rule_interval: 5
+    flash.proxy.config: "%s/conf/tiflash-learner.toml"
+    status.metrics_port: %d
+    logger.errorlog: "%s/log/tiflash_error.log"
+    logger.log: "%s/log/tiflash.log"
```
Comment (AstroProfundis) [inline pkg/meta/logic.go:615]: ditto

File: pkg/meta/logic.go [hunk 9]
```diff
@@ -526,6 +544,219 @@ func (i *PDInstance) ScaleConfig(e executor.TiOpsExecutor, b *Specification, clu
...
+}
+
+// InitTiFlashLearnerConfig initializes TiFlash learner config file
+func (i *TiFlashInstance) InitTiFlashLearnerConfig(cfg *scripts.TiFlashScript, src map[string]interface{}) (map[string]interface{}, error) {
+	topo := TopologySpecification{}
+
+	err := yaml.Unmarshal([]byte(fmt.Sprintf(`
+server_configs:
+  tiflash-learner:
+    log-file: "%s/log/tiflash_tikv.log"
```
Comment (AstroProfundis) [inline pkg/meta/logic.go:663]: ditto

File: pkg/meta/logic.go [hunk 10]
```diff
@@ -526,6 +544,219 @@ func (i *PDInstance) ScaleConfig(e executor.TiOpsExecutor, b *Specification, clu
...
+    server.engine-addr: "%s:%d"
+    server.addr: "0.0.0.0:%d"
+    server.advertise-addr: "%s:%d"
+    server.status-addr: "%s:%d"
+    storage.data-dir: "%s/tiflash/data/tiflash"
```
Comment (AstroProfundis) [inline pkg/meta/logic.go:668]: ditto

File: pkg/template/scripts/tiflash.go [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+	"io/ioutil"
+	"os"
+	"path"
+	"text/template"
+
+	"github.com/pingcap-incubator/tiup/pkg/localdata"
+)
+
+// TiFlashScript represent the data to generate TiFlash config
+type TiFlashScript struct {
```
Comment (AstroProfundis) [inline pkg/template/scripts/tiflash.go:27]: `DataDir` and `LogDir` are not used in the startup script, but in the configuration template hard coded in `logic.go`, maybe we could add dedicated templates for the config files?

File: pkg/meta/logic.go (hunk 4 above)
Comment (AstroProfundis) [reply pkg/meta/logic.go:673]: Um, I believe `Role()` is still needed when importing

File: pkg/meta/topology.go (hunk 5 above)
Comment (AstroProfundis) [reply pkg/meta/topology.go:320]: Um, I believe `Role()` is still needed when importing

Comment (lonng) [reply pkg/meta/topology.go:320]: ok

File: templates/scripts/run_tiflash.sh.tpl [hunk 12]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+set -e
+ulimit -n 1000000
```
Comment (lonng) [inline templates/scripts/run_tiflash.sh.tpl:3]: Prefer to specify the max open files in systems service scripts.

File: pkg/meta/logic.go (hunk 6 above)
Comment (birdstorm) [reply pkg/meta/logic.go]: done

File: pkg/meta/logic.go (hunk 7 above)
Comment (birdstorm) [reply pkg/meta/logic.go]: done

File: pkg/meta/logic.go (hunk 8 above)
Comment (birdstorm) [reply pkg/meta/logic.go]: done

File: pkg/meta/logic.go (hunk 9 above)
Comment (birdstorm) [reply pkg/meta/logic.go]: done

File: pkg/meta/logic.go (hunk 10 above)
Comment (birdstorm) [reply pkg/meta/logic.go]: done

File: templates/scripts/run_tiflash.sh.tpl (hunk 12 above)
Comment (birdstorm) [reply templates/scripts/run_tiflash.sh.tpl]: done

**Pull request merged**